gabrielgomesml/training_project_server
# Feature/Users-crud

## What I did

- Create base of the project;
- Create Users model, repository, controller and routes.

## How to test

- First of all make sure you have the docker installed on your machine
- Open the project and use the command `nvm use`
- Run a `yarn` or `yarn install`
- Create a .env file in the root of the project (i will send the model by email)
- Run the command `docker-compose up -d` to create a docker container with the postgres database
- After that run the command `yarn typeorm migration:run` or `yarn migration` to update the database with migrations and `yarn start` for run the project
- Open insomnia, Postman or other similar softwares and try to make a GET request to the API (you can use /users route). You should received a 401 response status with two possible messages: 'You are not authenticated' or 'Incorrect token'
- After that, create an user (JSON example bellow) and use the email and password in a POST to '/user-auth' route. Try with the wrong and right password
- You'll receive a token to allow you to make all requests. In Insomnia you go in Auth > Bearer and past the token in their field. Now you can test the GET request to '/users' again, but you need to be fast, token expires in 60s (check this expiration)
- Now, change the expiration time in Controllers > AuthController > Line 35 to 1800 and try to make another requests to the API

Examples:

POST localhost:3001/users

```
{
	"email": "gabrielgomes@gmail.com",
	"password": "training",
	"name": "Gabriel",
	"surname": "Gomes",
	"photo_address": "https://upload.wikimedia.org/wikipedia/commons/9/95/Alex_Turner%2C_Way_Out_West_2018.jpg",
	"phone": "(81) 98562-7583",
	"role": 0,
	"active": true
}
```

GET localhost:3001/users

GET localhost:3001/users/:id

PATCH localhost:3001/users

```
{
	"surname": "Melo"
}
```

DELETE localhost:3001/users/:id

## How to run tests

- First make sure to have an user created in your database and a token with a good expiration time setted
- Then go to `'./src/__tests__/Users.test.ts'` and put and existing user Id in line 5, and an existing account and password in lines 6 and 7
- In a terminal run `yarn test`
